packages/compiler/src/compiler/scope.ts

```diff
@@ -15,65 +15,54 @@ export default class Scope {
+   * Every scope has its own local pointers that contains the indexes of the variables in the global stash.
    */
-  private static stash: unknown[] = [] // Stash of every variable value in every scope
-  private static readFromStash(index: number): unknown {
-    return Scope.stash[index]
-  }
-  private static addToStash(value: unknown): number {
-    Scope.stash.push(value)
-    return Scope.stash.length - 1
-  }
+  private stash: unknown[] = [] // Stash of every variable value in the global scope
+  private pointers: Record<string, number> = {} // Index of every variable in the stash in the current scope
 
-  private static modifyStash(index: number, value: unknown): void {
-    Scope.stash[index] = value
+  constructor(globalScope: Record<string, unknown> = {}) {
+    for (const [key, value] of Object.entries(globalScope)) {
+      this.pointers[key] = this.addToStash(value)
+    }
   }
 
-  /**
-   * Local scope
-   * Every scope has its own local stash that contains the indexes of the variables and constants
-   * in the global stash.
-   */
-  private consts: Record<string, number> = {} // Index of every constant in the stash in the current scope
-  private vars: Record<string, number> = {} // Index of every variable in the stash in the current scope
-
-  constructor() {}
+  private readFromStash(index: number): unknown {
+    return this.stash[index]
+  }
+  private addToStash(value: unknown): number {
+    this.stash.push(value)
+    return this.stash.length - 1
+  }
 
-  exists(name: string): boolean {
-    return name in this.consts || name in this.vars
+  private modifyStash(index: number, value: unknown): void {
+    this.stash[index] = value
   }
 
-  isConst(name: string): boolean {
-    return name in this.consts
+  exists(name: string): boolean {
+    return name in this.pointers
   }
 
   get(name: string): unknown {
-    const index = this.consts[name] ?? this.vars[name] ?? undefined
+    const index = this.pointers[name] ?? undefined
     if (index === undefined)
       throw new Error(`Variable '${name}' does not exist`)
-    return Scope.readFromStash(index)
-  }
-
-  defineConst(name: string, value: unknown): void {
-    if (this.exists(name)) throw new Error(`Variable '${name}' already exists`)
-    this.consts[name] = Scope.addToStash(value)
+    return this.readFromStash(index)
   }
 
   set(name: string, value: unknown): void {
-    if (this.isConst(name))
-      throw new Error(`Constant '${name}' cannot be modified`)
     if (!this.exists(name)) {
-      this.vars[name] = Scope.addToStash(value)
+      this.pointers[name] = this.addToStash(value)
       return
     }
-    const index = this.vars[name]!
-    Scope.modifyStash(index, value)
+    const index = this.pointers[name]!
+    this.modifyStash(index, value)
   }
 
   copy(): Scope {
     const scope = new Scope()
-    scope.consts = { ...this.consts }
-    scope.vars = { ...this.vars }
+    scope.pointers = { ...this.pointers }
     return scope
   }
 }
```

Comment:
why do we need to continue to support variables?

Reply:
also missing tests for this if we indeed need to continue to support it

Reply:
creating variables inside of `{{#if }}` and `{{#each}}` statements should be scoped imo. I don't know about tags though.

Reply:
my question was more about whether we need variables in the first place

Reply:
Well, following the line of latitude-sdk's powerful template engine, here you can not only condition or iterate over variables, but you can also create and modify at will.

Reply:
The compiler uses variables in several ways, even if the user does not define one themselves. For example:
- Parameters are variables
- When using `{#each}` blocks, both the iterator element and index key are scoped variables
- Any functions we want to add (such as 'parse', for example) are treated as scoped variables for simplicity.

In addition to all of this, users can create their own variables, which will work just in the same way.

Reply:
oh ok then 👌🏼